AzureIoTHub.agent.lib.nut

```diff
@@ -629,513 +604,842 @@ class AzureIoTHub {
             }
-        }
 
-        function disconnect() {
-            _connection.close();
-            _resetConnectionTables();
+            _fillTopics(_connStrParsed.DeviceId);
         }
 
-        // done cb params - err
-        function sendEvent(message, done = null) {
-            if ( !_isOpen(_sessions.event) ) {
-                if (done) done("Cannot send while disconnected.");
-            } else {
-                if ( !_isOpen(_senders.event) ) {
-                    _log("creating event sender");
-                    // add message to queue
-                    _msgQueue.push({"msg" : message, "cb" : done});
-                    // create a new sender & send message
-                    _openEventSender();
-                } else {
-                    _log("event sender open")
-                    _sendEvent(message, done);
-                }
-            }
-        }
-
-        // done is called everytime a message is available, params - error, data
-        function receive(done) {
-            _handlers.onEvent <- done;
+        // Opens a connection to Azure IoT Hub.
+        //
+        // Returns:                         Nothing.
+        function connect() {
+            if (_state == S_DISCONNECTED) {
+                _log("Connecting...");
 
-            if (done == null) {
-                _receivers.event.close();
-                // clear out handlers
-                return this;
-            }
+                local devPath = "/" + _connStrParsed.DeviceId;
+                local username = _connStrParsed.HostName + devPath + format("/api-version=%s", AZURE_API_VERSION);
+                local resourcePath = "/devices" + devPath + format("/api-version=%s", AZURE_API_VERSION);
+                local resourceUri = AzureIoTHub.Authorization.encodeUri(_connStrParsed.HostName + resourcePath);
+                local passwDeadTime = AzureIoTHub.Authorization.anHourFromNow();
+                local sas = AzureIoTHub.SharedAccessSignature.create(
+                    resourceUri, null, _connStrParsed.SharedAccessKey, passwDeadTime).toString();
 
-            if( !_isOpen(_sessions.event) ) {
-                _handlers.onEvent("Cannot receive while disconnected.", null);
-            } else {
-                // TODO: test/handle if receiver already open
-                if ( _isOpen(_receivers.event) ) _receivers.event.close();
-                _openEventReceiver();
-            }
-            return this;
-        }
-
-        // PRIVATE FUNCTIONS
-        // ------------------------------------------------------------------------------------
-        // ------------------------------------------------------------------------------------
-
-        // Open Auth/Event Session/Sender/Receivers
-        // ------------------------------------------------------------------------------------
+                local options = {
+                    "username" : username,
+                    "password" : sas
```

Comment:
`.toString()`?

Reply:
`sas` is a string. The code above: `local sas = ....toString();`